### PR TITLE
New notification system: implementation

### LINKED ISSUE #10917: Notifications: implement the new system
I got feedback and review on https://github.com/readthedocs/readthedocs.org/pull/10890 and it seems we are pretty aligned with the design. The next step is the implementation of it.

I'm creating this issue here to track this work and discuss whatever is needed.

* Closes: https://github.com/readthedocs/readthedocs.org/issues/4226
* Closes: https://github.com/readthedocs/readthedocs.org/issues/3399
* Closes: https://github.com/readthedocs/readthedocs.org/issues/5909
* Closes: https://github.com/readthedocs/readthedocs.org/issues/9110
* Closes: https://github.com/readthedocs/readthedocs.org/issues/9279
* Closes: https://github.com/readthedocs/readthedocs.org/issues/4235

### PR DESCRIPTION
**Pull request by humitos**
Initial work towards the new notification system. I started adding some comments and deleting old code to understand the scope of the project.

### Decisions made

* Create an endpoint on APIv2 to be used from the builders. Builders use an API key to have write access on specific resources. We don't have this permission/authentication process implemented in APIv3 and there is no easy way to do it right now. Builders configure the API client using v2 also, so we should migrate all together. This is out of the scope of this project.
* The font-end will use APIv3 to list and update notifications following the new patterns.
* Some notification require "extra data" to render them properly (e.g. concurrency limit notification needs to render the actual limit number). These extra data is passed when the notification is created using `format_values` attribute and they are stored in a JSONField in the database, so they can be used to render the message at display time.
* Old admin actions to send notification to owners were removed to reduce complexity. We haven't been using them in a while now and they don't seem to be super useful since we have most of them automated now.
* Created a message "registry" so each application can register their own messages IDs to attach to the notification they send to users. This is done by creating a `<app>/notifications.py` and defining and registering the messages there. Then this file has to be imported from `<app>/apps.py` or `<app>/__init__.py` or similar approach.
* Use a custom `models.Manager` that exposes a `.add()` method with de-duplication capabilities.
* Hack old dashboard templates with a small `location.reload()` JS to avoid re-implementing the Knockout dynamic capabilities in the old dashboard.
* I copied&pasted the current messages and quickly adapted some of them. However, I'm expecting to improve them during the review process by getting feedback or even in a different PR to avoid blocking it.
* The work for "new messages" (that didn't exist before) won't be covered by this PR since I'm considering out-of-scope. For now, I'm sticking to the plan of implementing the new notification system and migrate the current notifications into the new system.
* `Build.error` field is not populated anymore from the builder. Those error are now attached as `Notification` objects to the build itself.

### ToDo

The work listed here is not mandatory of this PR, but I'm writing it down to keep track of it and don't forget:

- [ ] create a periodic Celery task to check for notifications that should be removed since the status has changed (e.g. subscription was paid, custom domain was verified, or similar). See https://github.com/readthedocs/readthedocs.org/issues/10981
- [ ] "cancel" notification when state changes via Django forms
- [x] implement the same `location.reload()` hack in the new dashboard until we write the proper Knockout code using APIv3
- [x] remove Django messages extend dependency in case it's not used anymore
- [x] remove `SiteNotification` class completely
- [x] implement this changes in `readthedocs-corporate` at https://github.com/readthedocs/readthedocs-corporate/pull/1699
- [ ] implement APIv3 to PATCH notifications and mark them as `read` for example
- [x] finish the migration of all the `ConfigError` notifications shown at build time 
- [ ] consider renaming `Notification.format_values` to `.context` which is more common in the Django world
- [x] implement this changes in `ext-theme` at https://github.com/readthedocs/ext-theme/pull/254
- [x] review messages from `readthedocs-ext` and implement the new notification system there as well
- [x] make all the old tests to pass
- [x] write down new test cases if needed (in case the old ones not cover current behavior)
- [ ] write a migration for `Project.skip=True` that create `Notification` object for those (see https://github.com/readthedocs/readthedocs.org/issues/10980)

### Notification error attached to a `Build` object

Screenshot_2023-12-21_12-24-17

### Notification error attached to a `Project` object

Screenshot_2023-12-21_12-24-09

### Notification error attached to a `User` object

Screenshot_2023-12-21_12-33-19

### Notification on business

Screenshot_2023-12-21_13-08-30

### Notification on new dashboard

Screenshot_2023-12-21_13-37-49

Closes #10917

### CODE REVIEW

File: readthedocs/api/v3/serializers.py [hunk 1]
```diff
@@ -59,10 +60,33 @@ class Meta:
         fields = []
 
 
+# TODO: decide whether or not include a `_links` field on the object
```
Comment (agjohnson) [inline readthedocs/api/v3/serializers.py:63]: Notifications probably don't need a detail view, but it would help to have a dismiss URL. Right now, we do this at the template load time:

https://github.com/readthedocs/ext-theme/blob/4bfe050e5f4db2855490a47e0524685c8bad33a2/readthedocsext/theme/templates/includes/utils/messages.html#L16

We won't be able to do this with the new notification system. It's best that the front end never have to recreate URL creation logic, so I think we do want some links returned on the notification API response.

File: readthedocs/api/v3/serializers.py [hunk 2]
```diff
@@ -59,10 +60,33 @@ class Meta:
...
+#
+# This also includes adding `/api/v3/notifications/<pk>` endpoint,
+# which I'm not sure it's useful at this point.
```
Comment (agjohnson) [inline readthedocs/api/v3/serializers.py:66]: Not sure I follow this comment though. Why is this required?

File: readthedocs/notifications/constants.py [hunk 3]
```diff
@@ -10,25 +8,20 @@
...
-    INFO: message_constants.INFO_PERSISTENT,
-    WARNING: message_constants.WARNING_PERSISTENT,
-    REQUIREMENT: message_constants.WARNING_PERSISTENT,
-    ERROR: message_constants.ERROR_PERSISTENT,
-}
+# TODO: use Enum here for these ones
+# Status
+UNREAD = "unread"
+READ = "read"
+DISMISSED = "dissmissed"
```
Comment (agjohnson) [inline readthedocs/notifications/constants.py:15]:
```suggestion
DISMISSED = "dismissed"
```

File: readthedocs/notifications/constants.py [hunk 4]
```diff
@@ -10,25 +8,20 @@
...
-NON_PERSISTENT_MESSAGE_LEVELS = (
-    DEBUG_NON_PERSISTENT,
-    INFO_NON_PERSISTENT,
-    SUCCESS_NON_PERSISTENT,
-    WARNING_NON_PERSISTENT,
-    ERROR_NON_PERSISTENT,
-)
+# Icons
+SOLID = "solid"
+DUOTONE = "duotone"
```
Comment (agjohnson) [inline readthedocs/notifications/constants.py:27]: I noticed this elsewhere, but I think we should probably just change the notification icon to a icon_classes. There are other classes used too, `fa-swap-opacity`/etc. In this case, there is no separate field for icon type, just:

```Notification.icon_classes = "fas fa-user-group fa-swap-opacity"```

File: readthedocs/notifications/messages.py [hunk 5]
```diff
@@ -0,0 +1,434 @@
...
+
+
+CONFIG_YAML_MESSAGES = []
+
+# TODO: review the copy of these notifications/messages on PR review and adapt them.
+# Most of them are copied from what we had in `readthedocs.doc_builder.exceptions`
+# and slightly adapted to have a header and a better body.
+BUILD_MESSAGES = [
+    Message(
+        id=BuildAppError.GENERIC_WITH_BUILD_ID,
```
Comment (agjohnson) [inline readthedocs/notifications/messages.py:65]: I was going to suggest this pattern on some of the other messages. It seems wise to reference a shared constant instead of a string here, or I could see the lookup ids going out of sync.

File: readthedocs/notifications/migrations/0001_initial.py [hunk 6]
```diff
@@ -0,0 +1,76 @@
...
+                    ),
+                ),
+                ("message_id", models.CharField(max_length=128)),
+                (
+                    "state",
+                    models.CharField(
+                        choices=[
+                            ("unread", "unread"),
+                            ("read", "read"),
+                            ("dissmissed", "dissmissed"),
```
Comment (agjohnson) [inline readthedocs/notifications/migrations/0001_initial.py:47]: Note, this is a typo

File: readthedocs/api/v3/serializers.py (hunk 1 above)
Comment (humitos) [reply readthedocs/api/v3/serializers.py:63]: Yes, ideally, all these URLs should be returned in the API response, as we are doing with other resources with the `_links` attribute: https://docs.readthedocs.io/en/stable/api/v3.html#project-details

Then, the frontend can just rely on them being always correct 👍🏼

File: readthedocs/api/v3/serializers.py (hunk 2 above)
Comment (humitos) [reply readthedocs/api/v3/serializers.py:66]: This comment seems to be incorrect here 😓. I remember writing this in the context of the `urls.py` file. I'm not sure what's doing here 🙃 

Re-reading it, maybe I'm referring here that since it adds the `_self` attribute pointing to `notifications-details` it will use `/api/v3/notifications/<pk>` 🤔

File: readthedocs/notifications/messages.py (hunk 5 above)
Comment (humitos) [reply readthedocs/notifications/messages.py:65]: Yeah, this is the correct pattern.

The places where I was using hard-coded strings are WIP and I plan to move them to this pattern. The IDs may be under a class defining an exception or just global constants.

**Comment by humitos** [comment]
@agjohnson thanks for jumping in with an initial review of this work!

**Comment by agjohnson** [comment]
Absolutely! I should also say that everything is looking great so far, nothing stood out as a problem. If you still need to throw around any ideas, let me know. I'm happy to talk through any of this. I suspect we might have some new work once we get to the front end pieces on the new dashboard, but I might try to do some initial attempts at this soon too.

**Comment by humitos** [comment]
Currently, I don't have any new ideas or similar. In fact, my plan is the following:

- review all the notes I wrote and decide whether or not keep/resolve them
- implement all the missing pieces to be able to _replace_ the old system with the new one _without_ adding any new feature in this PR
- update the test cases to pass
- keep the `location.reload()` hack/workaround for now and tackle that Knockout work in a following PR
- implement all the _new features_ in following PRs

I'm leaving some of this work for future PRs because this one touches a lot of files and it will be hard to keep it updated (out of conflict) with more working coming in the following weeks.

**Comment by humitos** [comment]
@agjohnson I'm marking this PR ready for review. I know it's _huge_, so, let me know if you want to do pair-reviewing 😄 . Feel free to ask any question here as well, tho.

I will be performing some extra QA locally as well to see if I find any obvious issue.

**Comment by humitos** [comment]
Lint is failing because the requirements have changed and they are always installed from `main` branch.

File: readthedocs/api/v3/serializers.py (hunk 1 above)
Comment (humitos) [reply readthedocs/api/v3/serializers.py:63]: I will make these changes in a following PR when we start using APIv3 from the front-end.

File: readthedocs/api/v2/serializers.py [hunk 7]
```diff
@@ -378,3 +382,53 @@ def get_username(self, obj):
...
+        try:
+            return self.queryset.get(pk=pk)
+        except (ObjectDoesNotExist, ValueError, TypeError):
+            self.fail("does_not_exist")
+
+
+class NotificationSerializer(serializers.ModelSerializer):
+    # Accept different object types (Project, Build, User, etc) depending on what the notification is attached to.
+    # The client has to send a value like "<model>/<pk>".
+    # Example: "build/3522" will attach the notification to the Build object with id 3522
```
Comment (agjohnson) [inline readthedocs/api/v2/serializers.py:423]: Does this mean to query notifications we'd use something like `/api/v2/notifications/?attached_to=project/123`?

For consistency of the API, we'd normally use `/api/v2/notifications/?project__pk=123`

I don't quite know if it's something we have control over though.

**Review by agjohnson — COMMENTED** [review]
Overall, this looks really good. It took a lot to review, so I'm sure it took a lot of energy to manage this change :sweat_smile:

I don't see anything that should hold us up for much longer. There are some changes that should come before this gets to release, just a number of minor pieces though.

There were also many comments for later and I don't want to lose these tasks or notes, so we should move these to issues. I commented on a few, here they are at a high level:

- [ ] API: add _links field to notification API response to pointing to URLs like the dismiss URL for the notification
- [ ] API: create an v3 endpoint to PATCH Build/Project notifications. This way the front-end can mark them as READ/DISMISSED
- [ ] API: create an v3 endpoint to list notifications for Projects
- [ ] Notifications: audit and tune notification body and subject copy
- [ ] Notifications: re-implement sending email notifications to users
- [ ] Tests: adapt/remove test_notification tests because we won't have backends anymore.

There were others too, but they seemed minor.

At a high level, I wanted to note two things:

- The one piece that feels missing currently is high priority messages -- messages that we want to show on the object detail views, but also on unrelated views. An example might be subscription failures, or major project events. Otherwise, the user has to navigate to these specific object's detail views to see the errors. This could make some errors undiscoverable. Perhaps this is going to be what the bell notification menu is used for, instead of just news.
- Organization notifications don't seem to be emit in any templates, which would orphan some of these notificaitons.

File: readthedocs/api/v3/views.py [hunk 8]
```diff
@@ -381,6 +384,32 @@ def create(self, request, **kwargs):  # pylint: disable=arguments-differ
...
+    # http://chibisov.github.io/drf-extensions/docs/#usage-with-generic-relations
+    def get_queryset(self):
+        queryset = self.queryset.filter(
+            attached_to_content_type=ContentType.objects.get_for_model(Build)
+        )
+
+        # TODO: make sure if this particular filter should be applied here or somewhere else.
+        return queryset.filter(
+            attached_to_id=self.kwargs.get("parent_lookup_build__id")
+        )
```
[the remark this answers is not in the document]
Comment (humitos) [reply readthedocs/api/v3/views.py:410]: I think we can be flexible here. We can have 1) one API endpoint per resource or 2) only one global endpoint with a filter per resource.

I started doing 1) as a test because is what DRF extensions allows to do and it's a pattern we have been following for other endpoints too. Also, I think it's easier to understand what notifications you are getting back. An example of the request would be: 

* Notifications for a Build: `/api/v3/project/docs/builds/123456/notifications/`
* Notifications for a Project: `/api/v3/project/docs/notifications/`
* etc..

The particular view where you commented (`NotificationsViewSet`) would cover 2) --it's not finished, and can probably be commented out/removed for now-- and would require the user sending the `?attached_to=build` and also `?attached_to_id=123456` so the endpoint knows what's the exact object to retrieve notifications for: `/api/v3/notifications/?attached_to=build&attached_to_id=123456`

I didn't finish this because I wanted to discuss a little more with you before making the final decision. I personally prefer 1) since it's a cleaner UX and it's already working 😄 . Do you have any preference here?

However, the view `NotificationsViewSet` may also be required for a different usage: _mark a specific notification object as read_ via the URL: `PATCH /api/v3/notifications/718273/`. So, we may keep it for that purpose.

We don't have to decide this _right now_, since it's outside the scope of this PR, tho.

File: readthedocs/config/notifications.py
[the remark this answers is not in the document]
Comment (humitos) [reply readthedocs/config/notifications.py:1]: Yes, I haven't change them. I may fixed a typo or similar but we definitely need to audit these messages since they haven't been super clear to users historically. Now, we have a lot more granular control and we can raise more particular messages for specific errors.

We have a bunch of reported issues related to this. I created a small PR to show how this work could be: https://github.com/readthedocs/readthedocs.org/pull/10979

I started working in another PR that instead of raising an error will attach a warning message (e.g. "Using `conda` key together with `build.tools.python: 3` has no effect. You can remove it", or similar) but it's not finished yet.

File: readthedocs/core/utils/contact.py [hunk 9]
```diff
@@ -4,71 +4,53 @@
...
 
-from readthedocs.notifications import SiteNotification
-from readthedocs.notifications.backends import SiteBackend
 
 log = structlog.get_logger(__name__)
 
 
+# TODO: re-implement sending notifications to users.
+# This needs more thinking because the notifications where "Generic",
+# and now we need to register a  ``Message`` with its ID.
```
[the remark this answers is not in the document]
Comment (humitos) [reply readthedocs/core/utils/contact.py:14]: This is a small utility script to send email _and_ site notifications to users. We used to contact them when deprecating stuffs and Santos originally created it for specific custom domain issues, IIRC.

I adapted it to keep sending email notifications, but I removed the site notifications for now because we need a `message_id` to store in the `Notification` object and the original sent messages are lost now --since they were temporal.

To keep using this script, we need to adapt it to:

1. receive a `message_id=` argument
2. create a new `Notification` object with the `message_id` passed
3. receive `attached_to` argument to know where to attach the notification to (e.g. `User`, `Project`, etc)

Then, the user/caller of this script will need to:

1. create a `Message` with the notification they want to send and register it using `registry.add(messages)`
2. call the script with this `message_id` and `attached_to`

This is required because now we are not storing the message into the database, but only the `message_id`. This means the message has to be _declared in the code_ at the moment of rendering the site notification to the user.

I will create an issue for this 👍🏼

File: readthedocs/notifications/messages.py [hunk 10]
```diff
@@ -0,0 +1,389 @@
...
+        header=_("Unknown problem"),
+        # Note the message receives the instance it's attached to
+        # and could be use it to inject related data
+        body=_(
+            """
+                There was a problem with Read the Docs while building your documentation.
+                Please try again later.
+                If this problem persists,
+                report this error to us with your build id ({instance.pk}).
+                """
```
[the remark this answers is not in the document]
Comment (humitos) [reply readthedocs/notifications/messages.py:79]: I suppose yes. The API response returns `\n    \n    Please try....` but that's not an issue because it's HTML. However, _I don't want that if possible_. I tried using `textwrap.wrap` and similar and I didn't find a good combination that removes all the trailing spaces and extra indentation.

Ideally, I'd like to:

- keep a clear indentation to make it easy to read the code
- don't use a big hundred chars "single-line" string
- don't use escape chars like `\` inside the string
- call the magic method (e.g. `textwrap.wrap`) only once inside the `registry.add()` method

Do you know a good way to do this?

File: readthedocs/notifications/messages.py [hunk 11]
```diff
@@ -0,0 +1,389 @@
...
+        ),
+        type=ERROR,
+    ),
+    Message(
+        id=BuildMaxConcurrencyError.LIMIT_REACHED,
+        header=_("Maximum concurrency limit reached."),
+        # TODO: how we are going to format variables (e.g. ``limit`` here)?
+        # The variables are passed when it's instantiated.
+        # However, we need to render the notification from the front-end in a different moment.
+        # Do we want to store the key/values in the database and use them to render the message?
```
[the remark this answers is not in the document]
Comment (humitos) [reply readthedocs/notifications/messages.py:101]: Nope, this problem is solved with `format_values=`. Thanks!

File: readthedocs/notifications/messages.py [hunk 12]
```diff
@@ -0,0 +1,389 @@
...
+        type=ERROR,
+    ),
+    Message(
+        id=BuildUserError.NO_CONFIG_FILE_DEPRECATED,
+        header=_("Your project doesn't have a <code>.readthedocs.yaml</code> file"),
+        body=_(
+            """
+                The configuration file required to build documentation is missing from your project.
+                Add a configuration file to your project to make it build successfully.
+                Read more at https://docs.readthedocs.io/en/stable/config-file/v2.html
```
[the remark this answers is not in the document]
Comment (humitos) [reply readthedocs/notifications/messages.py:226]: Yes. That makes sense to me 👍🏼

File: readthedocs/api/v2/serializers.py (hunk 7 above)
Comment (humitos) [reply readthedocs/api/v2/serializers.py:423]: This endpoint is on APIv2 and it's only used by the builders via `POST`, so we won't use for listing or similar.

This field, `attached_to` has a special syntax because the `Notification` is tied to a `ContentType`. The `attached_to` field communicates the resource _and_ the id of that resource in the same field with a special syntax: `<resource>/<id>`.

File: readthedocs/templates/builds/build_detail.html [hunk 13]
```diff
@@ -233,6 +245,20 @@ <h3>{% trans "Build Errors" %}</h3>
...
+      Show error notifications as "build-error" (e.g. "syntax error in YAML")
+
+      This temporal until we render these notifications from the front-end.
+    {% endcomment %}
+    {% for notification in build.notifications.all %}
+    {% if notification.get_message.type == "error" %}
+    <div class="build-error">
+        <h3>{{ notification.get_message.type|title }}</h3>
+        <p class="build-error">{{ notification.get_message.get_rendered_body|safe }}</p>
+    </div>
```
[the remark this answers is not in the document]
Comment (humitos) [reply readthedocs/templates/builds/build_detail.html:258]: Great!

File: readthedocs/templates/base.html [hunk 14]
```diff
@@ -103,8 +103,20 @@
...
+        {% endcomment %}
+        {% for notification in user_notifications %}
+        <p class="build-failure">
+            {{ notification.get_message.get_rendered_body|safe }}
+        </p>
+        {% endfor %}
+
         <!-- BEGIN notify -->
         {% block notify %}
+          {# TODO: migrate these notifications to new system #}
```
[the remark this answers is not in the document]
Comment (humitos) [reply readthedocs/templates/base.html:119]: There is no different plan for regular Django message like "Successfully logged out" and similar ones.

The "migration" required here is due that we were using `django-messages-extends` to add extra and custom site notifications and we are not using this pattern anymore since they are now `Notification` objects.

Note that "Close notification" button that marks the notification as read has to be migrated to the new system.

File: readthedocs/rtd_tests/tests/test_notifications.py [hunk 15]
```diff
@@ -1,55 +1,44 @@
+# TODO: adapt/remove these tests because we won't have backends anymore.
```
[the remark this answers is not in the document]
Comment (humitos) [reply readthedocs/rtd_tests/tests/test_notifications.py:1]: I ended up doing the work but it seems that I didn't remove the TODO comment 😄 . The invalid tests were removed and the remaining ones are passing.

File: readthedocs/notifications/messages.py [hunk 16]
```diff
@@ -0,0 +1,389 @@
...
+
+    def get_display_icon_classes(self):
+        if self.icon_classes:
+            return self.icon_classes
+
+        # Default classes that apply to all the notifications
+        classes = [
+            "fas",
+            "fa-user-group",
+            "fa-swap-opacity",
```
Comment (humitos) [inline readthedocs/notifications/messages.py:41]:
```suggestion
```

File: readthedocs/notifications/messages.py [hunk 17]
```diff
@@ -0,0 +1,389 @@
...
+        type=ERROR,
+    ),
+    Message(
+        id=MkDocsYAMLParseError.INVALID_DOCS_DIR_CONFIG,
+        header=_(""),
+        body=_(
+            """
+    The "docs_dir" config from your MkDocs YAML config file has to be a
+    string with relative or absolute path.
+                """
```
[the remark this answers is not in the document]
Comment (humitos) [reply readthedocs/notifications/messages.py:304]: We are discussing this at https://github.com/readthedocs/readthedocs.org/pull/10922#discussion_r1440310605

File: readthedocs/templates/core/project_bar_base.html [hunk 18]
```diff
@@ -29,16 +29,17 @@ <h1>
...
-            Your project is currently disabled for abuse of the system.
-            Please make sure it isn't using unreasonable amounts of resources or triggering lots of builds in a short amount of time.
-            Please <a href="mailto:{{ SUPPORT_EMAIL }}">contact support</a> to get your project re-enabled.
-          {% endblocktrans %}
-      </p>
-      <br>
-    {% endif %}
+    {% comment %}
+        All projects with "skip=True" needs to be attached a notification using the new system.
+        This notification will be shown here using the same style we used before.
```
[the remark this answers is not in the document]
Comment (humitos) [reply readthedocs/templates/core/project_bar_base.html:34]: I opened https://github.com/readthedocs/readthedocs.org/issues/10980 for this.

File: readthedocs/api/v3/views.py (hunk 8 above)
[the remark this answers is not in the document]
Comment (humitos) [reply readthedocs/api/v3/views.py:410]: I opened an issue for this at https://github.com/readthedocs/readthedocs.org/issues/10984

File: readthedocs/notifications/messages.py [hunk 19]
```diff
@@ -0,0 +1,389 @@
...
+        ),
+        type=ERROR,
+    ),
+    Message(
+        id=MkDocsYAMLParseError.INVALID_DOCS_DIR_PATH,
+        header=_(""),
+        body=_(
+            """
+                """
```
[the remark this answers is not in the document]
Comment (humitos) [reply readthedocs/notifications/messages.py:309]: I opened https://github.com/readthedocs/readthedocs.org/issues/10985 for this.

File: readthedocs/notifications/messages.py [hunk 20]
```diff
@@ -0,0 +1,389 @@
...
+            """
+        ),
+        type=INFO,
+    ),
+]
+
+BUILD_MKDOCS_MESSAGES = [
+    Message(
+        id=MkDocsYAMLParseError.GENERIC_WITH_PARSE_EXCEPTION,
+        header=_(""),
```
[the remark this answers is not in the document]
Comment (humitos) [reply readthedocs/notifications/messages.py:285]: I opened https://github.com/readthedocs/readthedocs.org/issues/10985 for this.

File: readthedocs/core/utils/contact.py (hunk 9 above)
[the remark this answers is not in the document]
Comment (humitos) [reply readthedocs/core/utils/contact.py:14]: I opened https://github.com/readthedocs/readthedocs.org/issues/10986 to track this work.

File: readthedocs/templates/builds/build_detail.html [hunk 21]
```diff
@@ -162,6 +163,12 @@
         </div>
       {% endif %}
 
+     {% comment %}
+        NOTE: I'm not migrating "deprecated build image used" nor "deprecated config used"
+        since these notifications are not useful anymore.
+        The YAML config file is required now and it's not possible to use `build.image` anymore.
+        We should probably remove these blocks completely now.
```
[the remark this answers is not in the document]
Comment (humitos) [reply readthedocs/templates/builds/build_detail.html:170]: I removed these blocks.

File: readthedocs/templates/base.html [hunk 22]
```diff
@@ -103,8 +103,20 @@
     <div id="content">
       <div class="wrapper">
 
+        {% comment %}
+          ``user_notifications`` comes from a Context Processor.
+          We need to use a CustomUser to have access to "user.notifications"
+          See https://docs.djangoproject.com/en/4.2/ref/settings/#std-setting-AUTH_USER_MODEL
+        {% endcomment %}
+        {% for notification in user_notifications %}
+        <p class="build-failure">
```
[the remark this answers is not in the document]
Comment (humitos) [reply readthedocs/templates/base.html:112]: Good point! It was just a mistake from copy&paste from another template. In fact, I was wondering why the style rendered was different than before 😅 . Now, it looks exactly as previously 👍🏼

File: readthedocs/api/v3/urls.py [hunk 23]
```diff
@@ -63,13 +63,28 @@
...
+# The front-end will use this endpoint.
+# allows /api/v3/projects/pip/builds/1053/notifications/
+builds.register(
+    r"notifications",
+    NotificationsViewSet,
+    basename="project-builds-notifications",
+    parents_query_lookups=["project__slug", "build__id"],
+)
+
+# TODO: create an APIv3 endpoint to PATCH Build/Project notifications.
```
[the remark this answers is not in the document]
Comment (humitos) [reply readthedocs/api/v3/urls.py:83]: https://github.com/readthedocs/readthedocs.org/issues/10983

File: readthedocs/api/v3/urls.py [hunk 24]
```diff
@@ -63,13 +63,28 @@
...
+# This way the front-end can mark them as READ/DISMISSED.
+#
+# TODO: create an APIv3 endpoint to list notifications for Projects.
```
[the remark this answers is not in the document]
Comment (humitos) [reply readthedocs/api/v3/urls.py:86]: https://github.com/readthedocs/readthedocs.org/issues/10984

File: readthedocs/notifications/messages.py (hunk 10 above)
[the remark this answers is not in the document]
Comment (humitos) [reply readthedocs/notifications/messages.py:77]: I opened https://github.com/readthedocs/readthedocs.org/issues/10987 to continue the discussion there.

File: readthedocs/templates/base.html (hunk 14 above)
[the remark this answers is not in the document]
Comment (humitos) [reply readthedocs/templates/base.html:123]: I opened an issue to continue talking about this at https://github.com/readthedocs/readthedocs.org/issues/10988

**Comment by humitos** [comment]
Thanks _a lot_ for the detailed review. I accepted most of the suggestions and created all the issues that are out-of-scope from this PR to put in the roadmap and work in the near future while working together in the front-end pieces. I created the issues for them:

> API: add _links field to notification API response to pointing to URLs like the dismiss URL for the notification
 
https://github.com/readthedocs/readthedocs.org/issues/10982

> API: create an v3 endpoint to PATCH Build/Project notifications. This way the front-end can mark them as READ/DISMISSED

https://github.com/readthedocs/readthedocs.org/issues/10983

> API: create an v3 endpoint to list notifications for Projects

https://github.com/readthedocs/readthedocs.org/issues/10984

> Notifications: audit and tune notification body and subject copy

https://github.com/readthedocs/readthedocs.org/issues/10985

> Notifications: re-implement sending email notifications to users

https://github.com/readthedocs/readthedocs.org/issues/10986

> Tests: adapt/remove test_notification tests because we won't have backends anymore.

There is no need to track this, since it was just that I forget to remove the TODO comment.

**Comment by humitos** [comment]
I'm merging this PR to avoid conflicts. I'll make any required work in following PRs.

**Comment by humitos** [comment]
We decided to merge this after today's deploy since I wasn't able to check everything _before_ the deploy.

I created the script to migrate the old `build.error` to the new `Notification` objects at https://github.com/readthedocs/readthedocs.org/issues/10980#issuecomment-1875562781. We can run it in production after deploying this PR.

**Comment by agjohnson** [comment]
> I think these notifications have to be attached to the correct object 

Definitely, I agree with this modeling and was instead thinking we'd use a separate query like `Notification.objects.filter(attached_to=ANY, news=True)`, or rename the `news` field to make more sense. I'm guessing this is only really a boolean, we don't see us needing a 3rd or more level of priority.

> show notifications attached to Organization to owners in all the pages

This seems like a good first step, and solves the more immediate problem of the organization notifications. We can always change these messages later to be flagged important, and use a more general pattern like a `important=True` flag.

> show notifications attached to Project to admins in project detail's page

:+1: I also can't think of any messages that are important yet immediately. User notifications are special cased though, and could use this same pattern instead of a dedicated query/template context.

Oh also, I realized we can maybe even show project error notifications in project listings, using an error icon for any projects with error notifications. That would at least give some higher level view into projects that have problems. 

> create a notification page that list all the notifications

I don't recall if we've talked about this, but it could be a great UI to add. I'd say this is maybe a later phase of work, it's nice to have but not essential to the current refactor.

File: readthedocs/projects/notifications.py [hunk 25]
```diff
@@ -1,28 +1,116 @@
...
+            """
+        ),
+        type=ERROR,
+    ),
+    Message(
+        id=RepositoryError.FAILED_TO_CHECKOUT,
+        header=_("Error while checking out the repository"),
+        body=_(
+            """
+            Failed to checkout revision: <code>{revision}</code>
```
Comment (stsewd) [inline readthedocs/projects/notifications.py:54]: Hmm, are we including these messages directly in the HTML? They could probably be abused to inject malicious content, since we are using `.format` to render the body, as opposed to something like Django's template engine that escapes HTML by default.

**Review by stsewd — COMMENTED** [review]
Did a quick look. The main concern is that we are rendering the notifications as `| safe`, but they include user content, that can be abused.

**Comment by agjohnson** [comment]
> I created the script to migrate the old build.error to the new Notification objects

Noted in chat, but there were some issues with the notification migration and it sounds like we're probably leaning towards skipping this. The template code and KO view would stay the same as it is currently, or wouldn't change much if we supported the list of notifications with the same logic.

File: readthedocs/api/v3/views.py (hunk 8 above)
[the remark this answers is not in the document]
Comment (agjohnson) [reply readthedocs/api/v3/views.py:410]: I'll follow along that issue, but to note here:

>  I personally prefer 1) since it's a cleaner UX and it's already working

I think I prefer this too. It doesn't make much difference to the front end as we need to resolve the API URL and pass that into the JS view either way.

> However, the view NotificationsViewSet may also be required for a different usage

Good point. On the front end, the API should give a URL for dismissing the notification in the API response, so the JS doesn't have to reconstruct this. So either way works here too, and it wouldn't be any more/less work to PATCH a deeper URL like `/api/v3/project/docs/builds/123456/notifications/789`

File: readthedocs/notifications/messages.py (hunk 10 above)
[the remark this answers is not in the document]
Comment (agjohnson) [reply readthedocs/notifications/messages.py:77]: Mentioned in #10987 already, I was going to suggest `dedent`. This does require wrapping each string still though. There may be some gettext internals that could be pieced together, but that seems hard and being redundant is explicit and easy.

We can continue on #10987 though, I don't think this needs to hold up merging. However, it might be best to avoid altering our translation files until these strings are resolved. I don't quite know how gettext handles moved strings vs _moved and altered_ strings, but I suspect the additional whitespace could throw off all the existing translations.

File: readthedocs/templates/base.html (hunk 14 above)
[the remark this answers is not in the document]
Comment (agjohnson) [reply readthedocs/templates/base.html:123]: > The "migration" required here is due that we were using django-messages-extends to add extra and custom site notifications and we are not using this pattern anymore since they are now Notification objects.

Yeah, I guess that is my confusion. The notifications remaining should be using Django messages and won't need to be ported. These will be mostly form confirmation and other temporary messages that don't require dismissal. This template block is already accurate then, but the conditionals around `message.pk` will never be truthy again and can be removed.

File: readthedocs/config/notifications.py
[the remark this answers is not in the document]
Comment (agjohnson) [reply readthedocs/config/notifications.py:1]: Yeah, I'm really excited for expanding notifications and more helpful notifications! This seems to fall under #10985, we can continue there.

**Comment by agjohnson** [comment]
> created all the issues that are out-of-scope from this PR to put in the roadmap and work in the near future

Thanks for creating all of those! I've added them to our roadmap and put the API changes on this sprint. The API pieces will be required to make progress on any front end pieces and can happen before we get to any HTML/JS.

I opened up readthedocs/ext-theme#259 to discuss how to proceed on the actual front end work, but I'm still feeling that these will be relatively minor compared to the last of the work on the backend.

- readthedocs/ext-theme#259

File: readthedocs/projects/notifications.py (hunk 25 above)
Comment (humitos) [reply readthedocs/projects/notifications.py:54]: Yes. Including the text directly in HTML was one of the initial requirements for this project.

Currently, we are not, and we _should not_ in the future, format data that comes from the user. All the data formatted comes from the backend and should be safe to use inside `.format` since we have full control of it.

Let me know if you found a particular case that you are worried about and we can take a deeper look at it.

File: readthedocs/notifications/models.py [hunk 26]
```diff
@@ -0,0 +1,76 @@
...
+    # UNREAD: the notification was not shown to the user
+    # READ: the notifiation was shown
+    # DISMISSED: the notification was shown and the user dismissed it
+    # CANCELLED: removed automatically because the user has done the action required (e.g. paid the subscription)
+    state = models.CharField(
+        choices=[
+            (UNREAD, UNREAD),
+            (READ, READ),
+            (DISMISSED, DISMISSED),
+            (CANCELLED, CANCELLED),
```
[the remark this answers is not in the document]
Comment (humitos) [reply readthedocs/notifications/models.py:32]: I don't have a strong preference. Either one is fine I think.

_Resolved_ conveys the "problem was solved" where _cancelled_ doesn't. So, 🤷🏼

**Comment by humitos** [comment]
I applied all the suggestions. I'm merging this so other people can start testing it locally in their day to day development. Please, let me know if anything doesn't work as you expected and I will try to fix it before deploying it next week.

File: readthedocs/projects/notifications.py (hunk 25 above)
Comment (stsewd) [reply readthedocs/projects/notifications.py:62]: Well, in this case the name of the revision comes from the user, an easier one to inject is from the config file

https://github.com/readthedocs/readthedocs.org/blob/e7b83e0c9a5da57236650843bda8025d7470c189/readthedocs/config/notifications.py#L108


```yaml
version: 2

<strong>hello</strong>: world
```

We should escape for HTML all values, otherwise this is vulnerable to stored XSS.

Comment (humitos) [reply readthedocs/projects/notifications.py:62]: PR at https://github.com/readthedocs/readthedocs.org/pull/10996
